geomesa-core/src/main/scala/org/locationtech/geomesa/core/index/IndexValueEncoder.scala

```diff
@@ -20,20 +20,156 @@ import java.nio.ByteBuffer
 import java.util.{Date, UUID}
 
 import com.vividsolutions.jts.geom.Geometry
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder
 import org.locationtech.geomesa.core
+import org.locationtech.geomesa.feature.{KryoFeatureEncoder, ScalaSimpleFeature, SimpleFeatureDecoder, SimpleFeatureEncoder}
+import org.locationtech.geomesa.utils.cache.SoftThreadLocalCache
 import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes._
 import org.locationtech.geomesa.utils.text.WKBUtils
+import org.opengis.feature.`type`.AttributeDescriptor
 import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
 
+import scala.collection.mutable
+
+/**
+ * Encode and decode index values
+ */
+trait IndexValueEncoder {
+
+  /**
+   * Fields that will be encoded/decoded
+   *
+   * @return
+   */
+  def fields: Seq[String]
+
+  /**
+   * Encodes a simple feature into a byte array. Only the attributes marked for inclusion get encoded.
+   *
+   * @param sf
+   * @return
+   */
+  def encode(sf: SimpleFeature): Array[Byte]
+
+  /**
+   * Decodes a byte array into a simple feature
+   *
+   * @param value
+   * @return
+   */
+  def decode(value: Array[Byte]): SimpleFeature
+
+}
+
+object IndexValueEncoder {
+
+  import org.locationtech.geomesa.utils.geotools.RichAttributeDescriptors.RichAttributeDescriptor
+
+  import scala.collection.JavaConversions._
+
+  private val cache = new SoftThreadLocalCache[String, (SimpleFeatureType, Seq[String])]()
+
+  def apply(sft: SimpleFeatureType, version: Int): IndexValueEncoder = apply(sft, None, version)
+
+  def apply(sft: SimpleFeatureType, transform: SimpleFeatureType, version: Int): IndexValueEncoder =
+    apply(sft, Some(transform), version)
+
+  def apply(sft: SimpleFeatureType, transform: Option[SimpleFeatureType], version: Int): IndexValueEncoder = {
+    val key = {
+      // cache key should be unique per feature - we use name + attributes + bindings
+      val descriptors = sft.getAttributeDescriptors.map(d => s"${d.getLocalName}:${d.getType.getBinding}}")
```

Comment:
Originally it was similar to that, but we ran into issues b/c of re-used sft names.
The user data doesn't get persisted right now, so we'd have to add that to SimpleFeatureTypes... I think I prefer keeping it localized to here.